packages/terminal/src/browser/terminal-widget-impl.ts

```diff
@@ -292,12 +293,12 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
         };
     }
 
-    get cwd(): Promise<URI> {
+    get cwd(): MaybePromise<URI> {
         if (!IBaseTerminalServer.validateId(this.terminalId)) {
             return Promise.reject(new Error('terminal is not started'));
         }
         if (this.terminalService.getById(this.id)) {
-            return this.shellTerminalServer.getCwdURI(this.terminalId)
+            return this.lastCwd || this.shellTerminalServer.getCwdURI(this.terminalId)
```

Comment:
In case the backend knows something we don't shall we try that first, and then fall back to last known if that fails (and maybe set it, if it succeeds)